src/core/or/connection_edge.c

```diff
@@ -883,14 +899,20 @@ export_hs_client_circuit_id(edge_connection_t *edge_conn,
   }
 
   /* Build the string */
-  tor_asprintf(&buf, "PROXY TCP6 %s:%x:%x %s %d %d\r\n",
+  tor_asprintf(&buf, "PROXY TCP6 %s:%s:%04x:%04x:%04x %s %d %d\r\n",
                src_ipv6_prefix,
-               gid >> 16, gid & 0x0000ffff,
+               rp_identity_buf ? rp_identity_buf : "0000:0000:0000",
+               get_options()->HiddenServiceExportInstanceID,
```

Comment:
This was actually the one I was most concerned about :-)
```suggestion
               (uint16_t) (get_options()->HiddenServiceExportInstanceID),
```